### ISSUE TITLE
Implement builtin template helper to dump available template variables

### ISSUE THREAD

**Issue by techtonik** [pallets/jinja#174]
Smarty template engine has a very convenient [{debug}](http://www.smarty.net/docs/en/language.function.debug.tpl) built-in that gives a quick overview of variables available for designer.

![Smarty_debug_console](https://f.cloud.github.com/assets/515889/74069/73886ee8-6077-11e2-96e3-41808672d2fe.jpg)

It would be nice to have the same stuff in Jinja2 [by default](http://stackoverflow.com/questions/3398850/how-to-get-a-list-of-current-variables-from-jinja-2-template)

**Comment by hvisage** [comment]
+1 :)

**Comment by Moini** [comment]
+1 from me, too (would like to have it available in pelican)

**Comment by westurner** [comment]
I'd have to check, but I think this should work:

```
<h2>locals()</h2>
<pre>{{ locals() }}</pre>
<h2>globals()</h2>
```

... https://stackoverflow.com/questions/3398850/how-to-get-a-list-of-current-variables-from-jinja-2-template

@mitsuhiko from https://stackoverflow.com/a/13757358/188833 : 

    import jinja2
    
    @jinja2.contextfunction
    def get_context(c):
	return c
    
    app.jinja_env.globals['context'] = get_context
    app.jinja_env.globals['callable'] = callable

for debugging, e.g. pyramid_debugtoolbar can show the template context if the view function -- appropriately abstracted IMHO -- just returns a context object (e.g. dict, OrderedDict)

* https://github.com/Pylons/pyramid_debugtoolbar
* https://westurner.org/wiki/awesome-python-testing#pyramid

## use case
My use case here is passing a few args from a CLI utility through to a jinja2 template (where I'd like to define defaults within the template (e.g. for Dockerfiles (without ``autoescape``, of all things))

What I'd like to do is

```jinja
{% set ENVVAR=context.get('ENVVAR', 'DEFAULT') %}
```

but, alas, the actual ``context`` object is [...]

**Comment by techtonik** [comment]
@westurner, so does it work? I am unable to check it right now, but it looks to me that `<pre>{{ locals() }}</pre>` should fail if it contains non-convertible bytes.

**Comment by westurner** [comment]
* ``json.dumps(locals())`` won't work without a custom JSONEncoder (with eg ``repr()`` for [code objects,])
* maybe ``pprint.pformat(locals())``? (this would be a convenient builtin)

**Comment by westurner** [comment]
* ... ``autoescape=True`` would then prevent XSS

**Comment by roytruelove** [comment]
+1

**Comment by exhuma** [comment]
This would be a boon to have when extending templates of a third-party app. I am currently using [hieroglyph](https://pypi.python.org/pypi/hieroglyph) to create slide-shows, which builds on sphinx, and thus Jinja2.

I've extended the default hieroglyph theme, and what I would like to find out is if the template contains some variable which tells me if I'm on a slide or not. Maybe the variable is there, maybe it isn't. So I am currently poking inside the hieroglyph source-code to find what I need.

Being able to simply dump all the available variables directly from *within* the template (that is, without modifying *any* Python code) would make this so much quicker and easier. So, let's say being able to do something like this would be nice:

```
{% extends "!layout.html %}

{# ... or any other sensible name #}
{% dump_variables() %}

{% block body %}
Hello World!
super()
Goodbye World
{% endblock %}
```

This way, I could introspect whatever hieroglyph offers me while developing the new theme.

**Comment by ShaheedHaque** [comment]
I created an extension for this purpose, it can be found at https://github.com/niwinz/django-jinja/files/1607805/jinja_extensions.py.txt

**Comment by techtonik** [comment]
@ShaheedHaque any screens?

**Comment by ShaheedHaque** [comment]
If you open the file, the comments have an example of the output. If
needed, I can provide an actual screenshot when I get home tonight...

**Comment by techtonik** [comment]
@ShaheedHaque screen would be really nice to compare with Smarty output from the first post.

**Comment by ShaheedHaque** [comment]
![image](https://user-images.githubusercontent.com/4881193/34664428-c23ad272-f453-11e7-8de2-9c1e44e37146.png)

As requested. You'll see that the contents are organised (and sorted) by context variables, filters and tests.

**Comment by techtonik** [comment]
Looks useful. So why not to fill a PR?

**Comment by ShaheedHaque** [comment]
I'd be happy to do that given some guidance on where in the ninja codebase the new code should live (I'm not familiar with this project).

**Comment by techtonik** [comment]
Me neither, but https://github.com/pallets/jinja/blob/master/jinja2/ext.py seems like a good start.

**Comment by ShaheedHaque** [comment]
See PR #798.

**Comment by davidism** [comment]
Will be available as a new extension in 2.11.

**Comment by westurner** [comment]
Thanks!

The debug extension docs should caution users about unintentional
disclosure of sensitive information
through the ``{% debug %}`` tag, and how to determine whether the extension
is enabled (so that a static analyzer can identify that someone forgot to
remove ``{% debug .* %}`` before release.
#983
